src/core/server/services/tenant/cache.ts

```diff
@@ -1,31 +1,61 @@
 import DataLoader from "dataloader";
 import { Redis } from "ioredis";
 import { Db } from "mongodb";
+import uuid from "uuid";
 
+import { EventEmitter } from "events";
+import logger from "talk-server/logger";
 import {
   retrieveAllTenants,
   retrieveManyTenants,
+  retrieveManyTenantsByDomain,
   Tenant,
 } from "talk-server/models/tenant";
 
-const CacheUpdateChannel = "tenant";
+const TENANT_UPDATE_CHANNEL = "tenant";
 
-// Cache provides an interface for retrieving tenant stored in local memory
-// rather than grabbing it from the database every single call.
-export default class Cache {
-  // private tenants: Map<string, Promise<Readonly<Tenant>>>;
-  private tenants: DataLoader<string, Readonly<Tenant> | null>;
-  private db: Db;
+const EMITTER_EVENT_NAME = "update";
+
+export type SubscribeCallback = (tenant: Tenant) => void;
+
+interface TenantUpdateMessage {
+  tenant: Tenant;
+  clientApplicationID: string;
+}
+
+// TenantCache provides an interface for retrieving tenant stored in local
+// memory rather than grabbing it from the database every single call.
+export default class TenantCache {
+  private tenantsByID: DataLoader<string, Readonly<Tenant> | null>;
+  private tenantsByDomain: DataLoader<string, Readonly<Tenant> | null>;
+  private mongo: Db;
+  private clientApplicationID: string;
+  private emitter: EventEmitter;
+
+  constructor(mongo: Db, subscriber: Redis) {
+    // Setup the event emitter.
+    this.emitter = new EventEmitter();
 
-  constructor(db: Db, subscriber: Redis) {
     // Save the Db reference.
-    this.db = db;
+    this.mongo = mongo;
+
+    // Create a new client application ID. This prevents duplicated messages
+    // generated by this application from being handled as external messages
+    // as we should have already processed it.
+    this.clientApplicationID = uuid.v4();
 
     // Prepare the list of all tenant's maintained by this instance.
-    this.tenants = new DataLoader(ids => retrieveManyTenants(db, ids));
+    this.tenantsByID = new DataLoader(ids => {
```

Comment:
Change this to  a _property intializer_?

```ts
tenantsByID = new DataLoader(ids => {
  logger.debug({ ids: ids.length }, "now loading tenants");
  return retrieveManyTenants(this.mongo, ids);
});
```

Reply:
Can you elaborate on this one?